File: singlyLinkedList.js
class Node{
    constructor(val){
    this.val = val;
    this.next = null
    }
}
  
  
class SinglyLinkedList{
    constructor(){
        this.head = null;
        this.tail = null;
        this.length = 0;
    }

    push(val){
     var newNode = new Node(val)
     if(!this.head){
       this.head = newNode;
       this.tail = this.head;
     }else{
       this.tail.next = newNode;
       this.tail = newNode;
     }
      this.length +=1;
      return this;
    }
    
    traverse(){
      let current = this.head
      while(current){
        console.log(current.val)
        current = current.next
      }
    }
    
    pop(){ //remove last node(tail)
      if(!this.head) return undefined;
      let current = this.head
      let newTail = current;
      while(current.next){
        newTail = current;
        current = current.next;
      }
        this.tail = newTail;
        this.tail.next = null
        this.length -= 1;
      if(this.length === 0){
        this.head = null
        this.tail = null
      }
        return current;
    }
    
    shift(){ //remove first node(head)
      if(!this.head) return undefined;
        let current = this.head;
        this.head = current.next;
        this.length -=1;
        return current;
    }
    
    unshift(val){ //adding a new node to head
      let newNode = new Node(val);
      if(!this.head){
        this.head = newNode;
        this.tail = this.head;
      }else{
        newNode.next = this.head;
        this.head = newNode;
      }
        this.length++;
        return this;
    }
    
    get(index){ //retrieving a node by it's position in the List
      if(index < 0 || index >= this.length) return null;
      let count = 0;
      let current = this.head
      while(count !== index){
        current = current.next;
        count++;
      }
      return current;
    }
    
    set(index, val){ //changing the value of a node based on it's position in the list
      let foundNode = this.get(index) 
      if(foundNode){
        foundNode.val = val
        return true;
      }
      return false;
    }
    
    insert(index, val){ //adding a node to the linked list at a specific position
      if(index === 0) return !!this.unshift(val);
      if(index < 0 || index > this.length) return false;
      if(index === this.length) return !!this.push(val);  // !! is a trick to return boolean for that is not boolean.
      let prevNode = this.get(index-1)
      let newNode = new Node(val)
      let nextNode = this.get(index)
      newNode.next = nextNode
      prevNode.next = newNode
      this.length++;
      return true;
    }
    
    remove(index){
      if(index === this.length-1) return this.pop();
      if(index === 0) return this.shift();
      if(index < 0 || index > this.length) return undefined;
      let prevNode = this.get(index-1)
      let removedNode = this.get(index)
      prevNode.next = removedNode.next
      this.length--;
      return removedNode;
    }
    
    reverse(){
    //head tail swap
        let node = this.head;
        this.head = this.tail;
        this.tail = node;

        let next;
        let prev = null;

        for(let i=0; i<this.length; i++){
            next = node.next;
            node.next = prev;
            prev = node;
            node = next;
        }
        return this;
    }
    
    print(){
        let arr = []
        let current = this.head
        while(current){
            arr.push(current.val)
            current = current.next
        }
        return arr
    }

  }
  
  var list = new SinglyLinkedList()
  list.push("1.Hello")
  list.push('2.man')
  list.push('3.nah')
  list.push('4.hm...')
  list.unshift('0.haha')
  
  console.log(list.print())
  list.reverse()
  console.log(list.print())
  // console.log(list.get(3))
  // console.log(list.set(1, 'shit'))
  // console.log(list.insert(2, 'new2'))
  // console.log(list.remove(1))
  // console.log(list.head.val)
  // console.log(list.head.next.val)
  // console.log(list.head.next.next.val)
  // console.log(list.head.next.next.next.val)
  // console.log(list.traverse())
  // console.log(list.pop())
  // console.log(list.shift())

File: data_structures/SinglyLinkedList.py
class Node:
    def __init__(self,val):
        self.val = val
        self.next = None


class SinglyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0

    def push(self,val):
        newNode = Node(val)
        if self.length == 0:
            self.head = newNode
            self.tail = self.head
        else:
            self.tail.next = newNode
            self.tail = newNode
        self.length += 1
        return self    

    def traverse(self):
        current = self.head
        while current:
            print(current.val)
            current = current.next 
        
    
    def pop(self):
        if not self.head:
            return None
        current = self.head
        newTail = current
        while current.next:
            newTail = current
            current = current.next
        self.tail = newTail
        self.tail.next = None
        self.length -= 1
        if self.length == 0:
            self.head = None
            self.tail = None
        return current

    def shift(self):
        if not self.head:
            return 'how do you return undefined.. in python?'
        current = self.head
        self.head = current.next
        self.length -= 1
        return current

    def unshift(self, val):
        newNode = Node(val)
        if not self.head:
            self.head = newNode
            self.tail = newNode
        else:
            newNode.next = self.head
            self.head = newNode
        self.length += 1
        return self

    def get(self, index):
        if index < 0 or index >= self.length:
            return None
        count = 0
        current = self.head
        while count is not index:
            current = current.next
            count += 1
        return current
    
    def set(self, index, val):
        foundNode = self.get(index)
        if foundNode:
            foundNode.val = val
            return True
        return False

    def insert(self, index, val):
        newNode = Node(val)
        oneBefore = self.get(index -1)
        oneAfter = self.get(index + 1)
        newNode.next = oneAfter
        oneBefore.next = newNode
        self.length += 1
        return True            

    def delete(self):
        if self.length == 0:
            return False
        elif self.length == 1:
            delNode = self.head.next
            self.head.next = None
            self.tail = self.head
            self.length -= 1
            return delNode.val
        else:
            delNode = self.head.next
            self.head.next = self.head.next.next
            self.length -= 1
            return delNode.val

    def search(self, val):
        check = self.head
        for i in range(self.length):
            if check.next.val == val:
                return Node
            check = check.next
        return None


    def reverse(self):
        node = self.head
        self.head = self.tail
        self.tail = node

        prev = None

        for i in range(self.length):
            next = node.next
            node.next = prev
            prev = node
            node = next
        return self
        


    
newList = SinglyLinkedList()
newList.push(5).push(10).push(15).push(20)
# print(newList.insert(1,99))
# print(newList.traverse())
print(newList.head.val)
print(newList.head.next.val)
print(newList.head.next.next.val)
print(newList.head.next.next.next.val)

newList.reverse()
print('------------------')
# print(newList.traverse())
# print(newList.pop().val)
# print(newList.pop().val)
# print(newList.pop().val)
# print(newList.set(3,9))
# print(newList.traverse())
# print(newList.head.next.next.val)
# newList.insert(5)
# newList.insert(3)
# newList.insert(99)
# print(newList.search(5))
